File: Hackerearth/Algorithms/Graphs/DFS/RhezoCritical.cpp
#include<iostream>
#include<bits/stdc++.h>
#define ll long long
using namespace std;
int idVal=0;
int countNodes=0;
void DFS(vector<int> ar[],int s,int parent,int visited[],int id[],int low[],ll sum[])
{
	id[s]=idVal++;
	low[s]=id[s];

	for(int i=0;i<(int)ar[s].size();i++)
	{
		int v=ar[s][i];
		if(v==parent)
			continue;
		if(visited[v]==0)
		{
			visited[v]=1;
			DFS(ar,v,s,visited,id,low,sum);
			
			low[s]=min(low[s],low[v]);
			sum[s]+=sum[v];
		}
		else
		{
			low[s]=min(low[s],id[v]);
		}
	}
}
void DFS2(vector<int> ar[],int s,int parent,int visited[],int id[],int low[],ll sum[],int p,int root)
{
	for(int i=0;i<(int)ar[s].size();i++)
	{
		int v=ar[s][i];
		if(v==parent)
			continue;
		if(visited[v]==0)
		{
			visited[v]=1;
			DFS2(ar,v,s,visited,id,low,sum,p,root);
			if(low[v]>id[s])
			{
				if(abs((sum[root]-sum[v])-sum[v])<p)
				{
					countNodes++;
				}
			}
		}
	}
}
void print(int ar[],int n)
{
	for(int i=0;i<n;i++)
	{
		cout<<ar[i]<<" ";
	}
	cout<<endl;
}
template<class t>
void init(t ar[],int ch,int n)
{
	for(int i=0;i<n;i++)
	{
		ar[i]=ch;
	}
}
int main()
{
	int n,m,p;
	cin>>n>>m>>p;
	vector<int> ar[n];
	for(int i=0;i<m;i++)
	{
		int a,b;
		cin>>a>>b;
		a--;
		b--;
		ar[a].push_back(b);
		ar[b].push_back(a);
	}

	int visited[n];
	int id[n];
	int low[n];
	ll sum[n];
	init<ll>(sum,1,n);
	init<int>(low,0,n);
	init<int>(id,0,n);
	init<int>(visited,0,n);

	//print(sum,n);

	for(int i=0;i<n;i++)
	{
		if(visited[i]==0)
		{
			//cout<<visited[i]<<endl;
			visited[i]=1;
			DFS(ar,i,i,visited,id,low,sum);
		}
	}

	init(visited,0,n);
	//print(id,n);
	//print(low,n);

	for(int i=0;i<n;i++)
	{
		if(visited[i]==0)
		{
			visited[i]=1;
			DFS2(ar,i,i,visited,id,low,sum,p,i);
		}
	}

	//print(sum,n);

	cout<<countNodes<<endl;

	return 0;
}

File: Hackerearth/Algorithms/Graphs/DFS/RhezoCritical.java
import java.io.*;
import java.util.*;
class Abc implements Runnable
{
	static class FastReader 
	{ 
	    BufferedReader br; 
	    StringTokenizer st; 
	
	    public FastReader() 
	    { 
	        br = new BufferedReader(new
	                 InputStreamReader(System.in)); 
	    } 
	
	    String next() 
	    { 
	        while (st == null || !st.hasMoreElements()) 
	        { 
	            try
	            { 
	                st = new StringTokenizer(br.readLine()); 
	            } 
	            catch (IOException  e) 
	            { 
	                e.printStackTrace(); 
	            } 
	        } 
	        return st.nextToken(); 
	    } 
	
	    int nextInt() 
	    { 
	        return Integer.parseInt(next()); 
	    } 
	
	    long nextLong() 
	    { 
	        return Long.parseLong(next()); 
	    } 
	
	    double nextDouble() 
	    { 
	        return Double.parseDouble(next()); 
	    } 
	
	    String nextLine() 
	    { 
	        String str = ""; 
	        try
	        { 
	            str = br.readLine(); 
	        } 
	        catch (IOException e) 
	        { 
	            e.printStackTrace(); 
	        } 
	        return str; 
	    } 
	} 
	public static void main(String arhs[])
	{
		new Thread(null, new Abc(), "Abc", 1<<26).start();
	}
	static int idVal=0;
	static int count=0;
	public static void DFS2(ArrayList<Integer> ar[],int s,int parent,int visited[],int low[],int id[],int sum[],int p,int root)
	{

		for(int i=0;i<ar[s].size();i++)
		{
			int v=ar[s].get(i);

			if(v==parent)
				continue;

			if(visited[v]==0)
			{
				visited[v]=1;
				DFS2(ar,v,s,visited,low,id,sum,p,root);
				if(low[v]>id[s])
				{
					if(sum[root]-sum[v]<p&&sum[v]<p)
					{
						count++;
					}
				}
			}
		}
	}
	public static void DFS(ArrayList<Integer> ar[],int s,int parent,int visited[],int low[],int id[],int sum[])
	{
		id[s]=idVal++;
		low[s]=id[s];
		
		for(int i=0;i<ar[s].size();i++)
		{
			int v=ar[s].get(i);
			
			if(v==parent)
				continue;
			
			if(visited[v]==0)
			{
				visited[v]=1;
				DFS(ar,v,s,visited,low,id,sum);
				low[s]=Math.min(low[s],low[v]);
			    sum[s]+=sum[v];
			}
			else
			{
			    low[s]=Math.min(low[s],id[v]);
			}
		}
	}
	public static void init(ArrayList<Integer> ar[])
	{
		for(int i=0;i<ar.length;i++)
		{
			ar[i]=new ArrayList<>();
		}
	}
	public void run()
	{
		FastReader in=new FastReader();
		int n,m,p;
		n=in.nextInt();
		m=in.nextInt();
		p=in.nextInt();
		ArrayList<Integer> ar[]=new ArrayList[n];
		init(ar);
		for(int i=0;i<m;i++)
		{
			int a,b;
			a=in.nextInt();
			b=in.nextInt();
			a--;
			b--;
			ar[a].add(b);
			ar[b].add(a);
		}
		
		int visited[]=new int[n];
		int low[]=new int[n];
		int id[]=new int[n];
		int sum[]=new int[n];
		
		for(int i=0;i<n;i++)
		{
			sum[i]=1;
		}

		for(int i=0;i<n;i++)
		{
			if(visited[i]==0)
			{
				visited[i]=1;
				DFS(ar,i,i,visited,low,id,sum);
			}
		}
		visited=new int[n];
		for(int i=0;i<n;i++)
		{
			if(visited[i]==0)
			{
				visited[i]=1;
				DFS2(ar,i,i,visited,low,id,sum,p,i);
			}
		}
		//System.out.println(Arrays.toString(sum));

		System.out.println(count);
	}
}